packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts

```diff
@@ -0,0 +1,248 @@
+  /**
+  * An optional Origin Access Control
+  * @default - an Origin Access Control will be created.
+  */
+  readonly originAccessControl?: cloudfront.IOriginAccessControl;
+
+  /**
+   * The level of permissions granted in the bucket policy and key policy (if applicable)
+   * to the CloudFront distribution.
+   * @default [AccessLevel.READ]
+   */
+  readonly originAccessLevels?: AccessLevel[];
+}
+
+/**
+ * Properties for configuring a S3 origin with OAI
+ */
+export interface S3BucketOriginWithOAIProps extends S3BucketOriginBaseProps {
+  /**
+  * An optional Origin Access Identity
+  * @default - an Origin Access Identity will be created.
+  */
+  readonly originAccessIdentity?: cloudfront.IOriginAccessIdentity;
+}
+
+/**
+ * A S3 Bucket Origin
+ */
+export abstract class S3BucketOrigin extends cloudfront.OriginBase {
+  /**
+   * Create a S3 Origin with Origin Access Control (OAC) configured
+   */
+  public static withOriginAccessControl(bucket: IBucket, props?: S3BucketOriginWithOACProps): cloudfront.IOrigin {
+    return new class extends S3BucketOrigin {
+      private originAccessControl?: cloudfront.IOriginAccessControl;
+
+      constructor() {
+        super(bucket, { ...props });
+        this.originAccessControl = props?.originAccessControl;
+      }
+
+      public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+        if (!this.originAccessControl) {
+          this.originAccessControl = new cloudfront.S3OriginAccessControl(scope, 'S3OriginAccessControl');
+        }
+
+        const distributionId = options.distributionId;
+        const accessLevels = new Set(props?.originAccessLevels ?? [cloudfront.AccessLevel.READ]);
+        const bucketPolicyActions = this.getBucketPolicyActions(accessLevels);
+        const bucketPolicyResult = this.grantDistributionAccessToBucket(distributionId!, bucketPolicyActions);
+
+        // Failed to update bucket policy, assume using imported bucket
+        if (!bucketPolicyResult.statementAdded) {
+          Annotations.of(scope).addWarningV2('@aws-cdk/aws-cloudfront-origins:updateImportedBucketPolicyOac',
+            'Cannot update bucket policy of an imported bucket. You will need to update the policy manually instead.\n' +
+            'See the "Setting up OAC with imported S3 buckets" section of module\'s README for more info.');
+        }
+
+        if (bucket.encryptionKey) {
+          const keyPolicyActions = this.getKeyPolicyActions(accessLevels);
+          const keyPolicyResult = this.grantDistributionAccessToKey(keyPolicyActions, bucket.encryptionKey);
+          // Failed to update key policy, assume using imported key
+          if (!keyPolicyResult.statementAdded) {
+            Annotations.of(scope).addWarningV2('@aws-cdk/aws-cloudfront-origins:updateImportedKeyPolicy',
+              'Cannot update key policy of an imported key. You may need to update the policy manually instead.');
```

Comment:
Is there some other plausible cause for why a user cant update key policy of an imported key? If so we should list it or just change this wording to say `You need to update...`

Reply:
I don't think so, I'll update the wording and point to the README section that describes the policy they need to add